tokio-timer/src/timer/level.rs

```diff
@@ -0,0 +1,201 @@
+use timer::{entry, Entry};
+
+use std::fmt;
+use std::sync::Arc;
+
+/// Wheel for a single level in the timer. This wheel contains 64 slots.
+pub struct Level {
+    level: usize,
+
+    /// Bit field tracking which slots currently contain entries.
+    ///
+    /// Using a bit field to track slots that contain entries allows avoiding a
+    /// scan to find entries. This field is updated when entries are added or
+    /// removed from a slot.
+    ///
+    /// The least-significant bit represents slot zero.
+    occupied: u64,
+
+    /// Slots
+    slot: [entry::Stack; LEVEL_MULT],
+}
+
+/// Indicates when a slot must be processed next.
+#[derive(Debug)]
+pub struct Expiration {
+    /// The level containing the slot.
+    pub level: usize,
+
+    /// The slot index.
+    pub slot: usize,
+
+    /// The instant at which the slot needs to be processed.
+    pub deadline: u64,
+}
+
+/// Level multiplier.
+///
+/// Being a power of 2 is very important.
+const LEVEL_MULT: usize = 64;
+
+impl Level {
+    pub fn new(level: usize) -> Level {
+        // Rust's derived implementations for arrays require that the value
+        // contained by the array be `Copy`. So, here we have to manually
+        // initialize every single slot.
+        macro_rules! s {
+            () => { entry::Stack::new() };
+        };
+
+        Level {
+            level,
+            occupied: 0,
+            slot: [
+                // It does not look like the necessary traits are
+                // derived for [T; 64].
+                s!(), s!(), s!(), s!(), s!(), s!(), s!(), s!(),
+                s!(), s!(), s!(), s!(), s!(), s!(), s!(), s!(),
+                s!(), s!(), s!(), s!(), s!(), s!(), s!(), s!(),
+                s!(), s!(), s!(), s!(), s!(), s!(), s!(), s!(),
+                s!(), s!(), s!(), s!(), s!(), s!(), s!(), s!(),
+                s!(), s!(), s!(), s!(), s!(), s!(), s!(), s!(),
+                s!(), s!(), s!(), s!(), s!(), s!(), s!(), s!(),
+                s!(), s!(), s!(), s!(), s!(), s!(), s!(), s!(),
+            ],
+        }
+    }
+
+    /// Finds the slot that needs to be processed next and returns the slot and
+    /// `Instant` at which this slot must be processed.
+    pub fn next_expiration(&self, now: u64) -> Option<Expiration> {
+        // Use the `occupied` bit field to get the index of the next slot that
+        // needs to be processed.
+        let slot = match self.next_occupied_slot(now) {
```

Comment:
I think this can just be
```rust
let slot = self.next_occupied_slot(now)?;
```

Reply:
Looks like that does not work on Rust 1.21.

Reply:
Ah, nevermind!